File: 84.java
import java.util.*;
import javafx.util.*;

class Solution {
    public int largestRectangleArea(int[] heights) {
        Deque<Pair<Integer, Integer>> monoStack = new ArrayDeque<>();
        int n = heights.length;
        int maxArea = 0;
        for (int i = 0; i < n; i++) {
            if (monoStack.isEmpty()) {
                monoStack.push(new Pair<Integer, Integer>(i, heights[i]));
                continue;
            }
            int lastPopIndex = i;
            while (!monoStack.isEmpty() && monoStack.peek().getValue() > heights[i]) {
                Pair<Integer, Integer> popped = monoStack.pop();
                lastPopIndex = popped.getKey();
                maxArea = Math.max(maxArea, popped.getValue() * (i - lastPopIndex));
            }
            monoStack.push(new Pair<Integer, Integer>(lastPopIndex, heights[i]));
        }

        while (!monoStack.isEmpty()) {
            Pair<Integer, Integer> popped = monoStack.pop();
            maxArea = Math.max(maxArea, popped.getValue() * (n - popped.getKey()));
        }
        return maxArea;
    }
}

File: Python/84.py
from typing import List


class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        max_area = 0
        stack = []  # (index, height)

        for i, h in enumerate(heights):
            start = i
            while stack and stack[-1][1] > h:
                index, height = stack.pop()
                max_area = max(max_area, height * (i - index))
                start = index
            stack.append((start, h))

        for i, h in stack:
            max_area = max(max_area, h * (len(heights) - i))

        return max_area
